### ISSUE TITLE
[BPF] Double-free when passing non-trivially-destructible ≤8B by-value parameters (e.g. std::unique_ptr)

### ISSUE THREAD

**Issue by choury** [llvm/llvm-project#207686]
### Summary

When compiling C++ for the BPF target, passing a **≤8B non-trivially-destructible type by value** (e.g. `std::unique_ptr`, 8 bytes) produces code that **double-frees**.

The BPF backend lowers such a parameter directly to `i64` (passing the pointer value by value in a register). However, Clang's frontend still emits a **backup temporary** on the caller side — move-constructed from the argument, then destroyed after the call — assuming the Itanium *invisible-reference* ABI where the callee would null the temp on move-out. Since the callee receives the value directly (not a pointer to the temp), it **cannot** null the temp, and the caller's destructor of the still-holding-pointer temporary double-frees.

No diagnostic is emitted; the miscompile is silent.

### Reproducer

Self-contained, no standard-library headers needed:

```cpp
struct U {                                  // like std::unique_ptr<int>: 8B, non-trivial dtor
    int* p;
    U();
    explicit U(int*);
    U(U&&) noexcept;                        // out-of-line, so the backup temp is not elided
    U& operator=(U&&) noexcept;
    ~U();
};
struct Node { int v; U child; };

__attribute__((noinline)) Node* mk(U l) {   // by-value non-trivial ≤8B param
    Node* n = new Node();
    n->v = 0;
    n->child = static_cast<U&&>(l);         // move param into member (callee takes ownership)
    return n;
}
int main(int argc, char**) {
    U leaf(new int(argc));
    Node* root = mk(static_cast<U&&>(leaf));
    return root->v;
}
// out-of-line definitions are opaque stubs (only needed for codegen; -emit-llvm doesn't require them)
U::U() : p(nullptr) {}
U::U(int* x) : p(x) {}
U::U(U&& o) noexcept : p(o.p) { o.p = nullptr; }
U& U::operator=(U&& o) noexcept { if (p) delete p; p = o.p; o.p = nullptr; return *this; }
U::~U() { if (p) delete p; }
```

### Command

```
clang++ -std=c++20 -target bpf -mcpu=v4 -O1 -S -emit-llvm -nostdinc++ repro.cpp -o repro.ll
```

### Resulting IR (BPF, buggy)

`mk`'s signature — the `U` parameter is lowered to `i64`:

```llvm
define dso_local ptr @_Z2mk1U(i64 %0) ...
```

In `main`, the caller materializes a backup temporary `%4`, move-constructs into it (`_ZN1UC1EOS_` = `U::U(U&&)`), then passes the pointer **value** (loaded from `%4` and `ptrtoint`'d) — *not* a pointer to `%4` — to `mk`. After the call, `mk` owns the pointer (it moved it into `e->child`), yet `main` destroys `%4`, which still holds that same pointer:

```llvm
  call void @_ZN1UC1EPi(ptr %3, ptr %5)                  ; construct leaf  (%3)
  call void @_ZN1UC1EOS_(ptr %4, ptr %3)                 ; move-construct backup %4 from leaf
  %6 = load ptr, ptr %4
  %7 = ptrtoint ptr %6 to i64
  invoke ptr @_Z2mk1U(i64 %7)                            ; mk takes ownership of the pointer
          to label %9 unwind label %11
9:
  call void @_ZN1UD1Ev(ptr %4)                           ; ~backup: deletes the same pointer → DOUBLE-FREE
  ...
  call void @_ZN1UD1Ev(ptr %3)                           ; ~leaf: noop (was nulled by the move-construct)
```

### Expected (x86_64 -O1, correct)

The same source compiled for x86_64 passes the temp **by pointer** (invisible-reference ABI). `mk` receives `ptr %1` (the caller's temp), reads the pointer, and **nulls the temp** on move-out — so the caller's destructor of the now-null temp is a noop:

```llvm
define void @_Z2mk1U(ptr sret %0, ptr noundef %1) ...
; inside mk:
  %5 = load ptr, ptr %1
  store ptr null, ptr %1          ; callee nulls the caller's temp
```

No double-free.

### Scope

Only **≤8B non-trivially-destructible by-value parameters** are affected (`std::unique_ptr`, any 8B type with a non-trivial destructor). Types **>8B** (`std::string` 24B, `std::shared_ptr` 16B) go through the `byval`/ptr invisible-reference path and are correct (callee shares the caller's temp and nulls it on move-out).

This is unrelated to the `byval` attribute: `unique_ptr` is non-trivially-copyable, so the frontend never emits `byval` for it. The mismatch is purely between Clang's frontend codegen (which assumes invisible-reference ABI for non-trivial by-value params) and the BPF backend's calling convention (which lowers ≤8B such params to `i64` by value).

### Versions affected

Reproduced on **clang 19.1.7, 20.1.8, 21.1.8, 22.1.8, and 23 trunk** (built 2026-07-05). The generated IR is byte-identical across all five versions — the bug has not changed in ~2 years and is not a recent regression.

### Suggested fix direction

Either:
- The BPF backend should use the invisible-reference ABI for non-trivially-destructible by-value params (pass `ptr` to a caller-owned temp, like x86), so the callee can null it on move-out; **or**
- Clang's frontend, when targeting BPF, should not emit the backup temporary + cleanup for params the backend will lower to `i64` by value; **or**
- At minimum, emit a diagnostic ("non-trivial by-value parameter not supported on BPF") instead of silently miscompiling — consistent with how the BPF backend already rejects `sret`/`byval`.

<details>
<summary>Full BPF IR (<code>repro.ll</code>, clang 19.1.7)</summary>

```llvm
; ModuleID = 'repro.cpp'
source_filename = "repro.cpp"
target datalayout = "e-m:e-p:64:64-i64:64-i128:128-n32:64-S128"
target triple = "bpf"

%struct.U = type { ptr }

define dso_local noundef nonnull ptr @_Z2mk1U(i64 %0) local_unnamed_addr #0 personality ptr @__gxx_personality_v0 {
  %2 = alloca %struct.U, align 8
  %3 = inttoptr i64 %0 to ptr
  store ptr %3, ptr %2, align 8
  %4 = tail call noalias noundef nonnull dereferenceable(16) ptr @_Znwm(i64 noundef 16) #8
  tail call void @llvm.memset.p0.i64(ptr noundef nonnull align 8 dereferenceable(16) %4, i8 0, i64 16, i1 false)
  %5 = getelementptr inbounds i8, ptr %4, i64 8
  invoke void @_ZN1UC1Ev(ptr noundef nonnull align 8 dereferenceable(8) %5)
          to label %6 unwind label %8
6:
  store i32 0, ptr %4, align 8
  %7 = call noundef nonnull align 8 dereferenceable(8) ptr @_ZN1UaSEOS_(ptr %5, ptr %2) #9
  ret ptr %4
8:
  %9 = landingpad { ptr, i32 } cleanup
  tail call void @_ZdlPvm(ptr noundef %4, i64 noundef 16) #10
  resume { ptr, i32 } %9
}

define dso_local noundef i32 @main(i32 noundef %0, ptr nocapture noundef readnone %1) local_unnamed_addr #6 personality ptr @__gxx_personality_v0 {
  %3 = alloca %struct.U, align 8
  %4 = alloca %struct.U, align 8
  call void @llvm.lifetime.start.p0(i64 8, ptr nonnull %3) #9
  %5 = tail call noalias noundef nonnull dereferenceable(4) ptr @_Znwm(i64 noundef 4) #8
  store i32 %0, ptr %5, align 4
  call void @_ZN1UC1EPi(ptr %3, ptr noundef nonnull %5)
  call void @_ZN1UC1EOS_(ptr %4, ptr %3) #9
  %6 = load ptr, ptr %4, align 8
  %7 = ptrtoint ptr %6 to i64
  invoke noundef ptr @_Z2mk1U(i64 %7)
          to label %9 unwind label %11
9:
  call void @_ZN1UD1Ev(ptr %4) #9
  %10 = load i32, ptr %8, align 8
  call void @_ZN1UD1Ev(ptr %3) #9
  call void @llvm.lifetime.end.p0(i64 8, ptr nonnull %3) #9
  ret i32 %10
11:
  %12 = landingpad { ptr, i32 } cleanup
  call void @_ZN1UD1Ev(ptr %4) #9
  call void @_ZN1UD1Ev(ptr %3) #9
  call void @llvm.lifetime.end.p0(i64 8, ptr nonnull %3) #9
  resume { ptr, i32 } %12
}
```
</details>

**Comment by choury** [comment]
Follow-up: the **Scope** section of this issue ("Types >8B go through the byval/ptr invisible-reference path and are correct") is **incorrect** — `std::shared_ptr<T>` (16 B) is also lowered by value, and also double-frees. The size cutoff in the current title (≤8B) understates the blast radius.

### Reproducer (16 B, no standard library)

```cpp
struct RC16 {              // same shape & semantics as std::shared_ptr<T>
    void* a; void* b;
    RC16();
    RC16(RC16&&);
    RC16(const RC16&);
    ~RC16();
};
void sink(RC16 p);
void caller(RC16 r) { sink(static_cast<RC16&&>(r)); }
```

```
clang++ -std=c++20 -target bpf -mcpu=v4 -O0 -S -emit-llvm rc16.cpp -o -
```

### Resulting IR (BPF, buggy)

```llvm
define dso_local void @_Z6caller4RC16([2 x i64] %0) #0 {
  %2 = alloca %struct.RC16, align 8
  %3 = alloca %struct.RC16, align 8
  store [2 x i64] %0, ptr %2, align 8
  call void @_ZN4RC16C1EOS_(ptr %3, ptr %2)            ; move-construct backup temp %3 from %2
  %4 = load [2 x i64], ptr %3, align 8
  call void @_Z4sink4RC16([2 x i64] %4)                ; sink takes the *value* (two i64s), not &%3
  call void @_ZN4RC16D1Ev(ptr %3)                      ; ~backup: destroys the same handle → DOUBLE-FREE
  ret void
}
```

`%3` is a backup temporary exactly like the ≤8 B `unique_ptr` case in the original report: move-constructed, the value handed to the callee as `[2 x i64]`, then destroyed. The callee cannot null `%3` (it never received a pointer to it). The only difference is that here the value is `2 × i64` instead of a single `i64`.

### Expected (x86_64 -O0, correct)

```llvm
define dso_local void @_Z6caller4RC16(ptr noundef %0) #0 {
  %2 = alloca %struct.RC16, align 8
  call void @_ZN4RC16C1EOS_(ptr %2, ptr %0)            ; move-construct temp from the *caller-owned* slot
  call void @_Z4sink4RC16(ptr %2)                       ; sink receives a pointer to the temp
  call void @_ZN4RC16D1Ev(ptr %2)                       ; sink has nulled the temp on move-out → noop
  ret void
}
```

Here `sink` shares the caller's temp by pointer and nulls it on move-out, so the caller's destructor is a noop.

### End-to-end crash with real `std::shared_ptr`

```cpp
std::shared_ptr<X> global;
__attribute__((noinline)) void store_in_global(std::shared_ptr<X> p) {
    global = std::move(p);     // callee takes ownership
}
int main() {
    auto p = std::make_shared<X>(42);
    store_in_global(std::move(p));
    return global->val;        // use-after-free
}
```

Running this (linked against musl + libc++, executed in our BPF VM, `clang-19 -O1`):

```
store: val=42 use_count=1
~X val=42                 ← object freed when the backup temp is destroyed after store_in_global returns
Memory access violation at PC 0x40001168: invalid read at address 0x40230060
```

`use_count` is `1` inside `store_in_global`, but `~X` runs as soon as `store_in_global` returns (the caller's destructor of the backup temp decrements the refcount to 0), so `global->val` reads freed memory. The mechanism is identical to the ≤8 B case in the original report.

### Versions

Byte-identical IR on **clang 19.1.7, 21.1.8, and 23 trunk** — `[2 x i64]` on BPF, `ptr` on x86_64 across all three. Not a regression.

### Implication

The `>8 B` cutoff in the original Scope came from the assumption that the BPF backend routes any aggregate that doesn't fit in a register through the indirect/invisible-reference path. That assumption is wrong: a 16 B non-trivially-destructible aggregate is still passed as `[2 x i64]` (two registers), so **every `std::shared_ptr<T>` (and any other >8 B non-trivially-destructible type that fits in two registers, e.g. a 16 B type with a non-trivial destructor)** is affected, not just `std::unique_ptr`. The general rule is "any non-trivially-destructible by-value parameter the BPF backend lowers to integer registers", not "≤8 B".

A backend-side fix (route all non-trivially-destructible by-value params through `ptr`, matching x86_64) would close both the ≤8 B and >8 B cases at once.
